migration/wrapper.go

```diff
@@ -0,0 +1,243 @@
+func NewMigrationRunnerWithHashFunction(hashFn func(string) (string, error)) *Runner {
+	return &Runner{hashFn}
+}
+
+// RunMigrations executes the migrations
+func (mr *Runner) RunMigrations(dbPath string) error {
+	repo, err := NewMigrationRepository(dbPath)
+	if err != nil {
+		return err
+	}
+
+	var bkpRepo *Repository
+	migrationList := make([]string, 0, len(migrations))
+	for k := range migrations {
+		migrationList = append(migrationList, k)
+	}
+	sort.Strings(migrationList)
+
+	//For each of them, in order execute
+	for _, k := range migrationList {
+		start := time.Now()
+		// Add hash check
+		if repo.Exists(k) {
+			log.Infof("Migration %s already run", k)
+			continue
+		}
+
+		// backup DB
+		bkpRepo, err = backupDB(repo, k)
+		if err != nil {
+			return nil
+		}
+
+		// execute migration file
+		if err = migrations[k](repo.db); err != nil {
+			log.Errorf("Migration %s failed", k)
+			err1 := revertDBToBackup(repo, bkpRepo)
+			if err1 != nil {
+				return err1
+			}
+			return err
+		}
+
+		// on success store migration item in DB
+		hash, err := mr.CalculateHash(k)
+		if err != nil {
+			err1 := revertDBToBackup(repo, bkpRepo)
+			if err1 != nil {
+				return err1
+			}
+			return err
+		}
+		mi := &Item{
+			ID:       k,
+			DateRun:  time.Now().UTC(),
+			Duration: time.Now().Sub(start),
+			Hash:     hash,
+		}
+		if err = repo.CreateMigration(mi); err != nil {
+			err1 := revertDBToBackup(repo, bkpRepo)
+			if err1 != nil {
+				return err1
+			}
+			return err
+		}
+
+		err = bkpRepo.db.Close()
+		if err != nil {
+			return err
+		}
+
+	}
+
+	return repo.db.Close()
+}
+
+func calculateMigrationHash(name string) (string, error) {
+	data, err := Asset(fmt.Sprintf("migration/files/%s.go", name))
+	if err != nil {
+		return "", err
+	}
+	hb, err := sha256Hash(data)
+	if err != nil {
+		return "", err
+	}
+	return hexutil.Encode(hb), nil
+}
+
+func getBackupName(path, name string) string {
+	bkpPath := strings.TrimSuffix(path, ".leveldb")
+	return fmt.Sprintf("%s_%s.leveldb", bkpPath, name)
+}
+
+func backupDB(srcRepo *Repository, migrationID string) (bkp *Repository, err error) {
+	//Closing src to make backup
+	err = srcRepo.db.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	dstPath := getBackupName(srcRepo.dbPath, migrationID)
+	err = CopyDir(srcRepo.dbPath, dstPath)
+	if err != nil {
+		return nil, err
+	}
+
+	// Refreshing src DB
+	err = srcRepo.RefreshDB()
+	if err != nil {
+		return nil, err
+	}
+
+	return NewMigrationRepository(dstPath)
+}
+
+func revertDBToBackup(srcDB, bkpDB *Repository) error {
+	erri := srcDB.db.Close()
+	if erri != nil {
+		return erri
+	}
+	erri = bkpDB.db.Close()
+	if erri != nil {
+		return erri
+	}
+
+	err := os.RemoveAll(srcDB.dbPath)
+	if err != nil {
+		return err
+	}
+	return os.Rename(bkpDB.dbPath, srcDB.dbPath)
+}
+
+// CopyFile copies a single file from src to dst
+// Code taken from https://blog.depado.eu/post/copy-files-and-directories-in-go
+func CopyFile(src, dst string) error {
```

Comment:
From here is copied code from source above. Maybe I could move it to a utils package, but at this point I do not see other use cases.